pkg/cache/internal/selector.go

```diff
@@ -0,0 +1,27 @@
+package internal
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// SelectorsByGVK associate a GroupVersionKind to a field/label selector
+type SelectorsByGVK map[schema.GroupVersionKind]Selector
+
+// Selector specify the label/field selector to fill in ListOptions
+type Selector struct {
```

Comment:
I like the the idea since user will be able to do other kind of matching reususe the client interface, but I will have to change stuff under `pkg/client/options` or wrap stuff under `pkg/cache` right ?

Reply:
If we would want to restrict which specific list options can be used for the cache selector, these options would need to add the `IsSelector` func, yes. It's not perfect, but would work.

Otherwise, we could just use
```go
type Selectors []client.ListOption
```
and only copy over field and label selectors in `ApplyToList`. But this would silently ignore "unsupported" list options (e.g. `client.Continue` or `client.Limit`), so I tend to prefer the other approach.

Maybe someone has a better idea than me :)

Reply:
Will add a new commit with the proposal to see how it rolls review wise, if it don't it will be easier to go back to previous proposal.

Reply:
@timebertt done at https://github.com/kubernetes-sigs/controller-runtime/pull/1435/commits/d5caf16870244174e4177a8d46241578571e4531

Reply:
That looks very promising, thanks!

One thing, I'm asking myself: should we also allow selecting a given namespace? Because currently one can only use either the `multiNamespaceCache` or the `informerCache`, but you can't restrict to a given namespace and by selector at the same time.
It would be easily achievable by also supporting `client.InNamespace` here, though. WDYT?

Reply:
I don' think using the clients list option is a great idea as the options there are a superset of what can be filtered for in the cache nor that passing client options into the cache is particularly intuitive.

Reply:
@qinqon

Reply:
the `pkg/cache` interface is going to use `client.Selctor` and not `client.ListOptions` so it's just a correct subset of ListOptions, they are not going to be able to pass for example a `client.ListOption{Namespace: "foo"}` since this is not implementing `IsSelector` and the ones that does are only selectors aware at "ApplyToList" 

```golang
type SelectorsByGVK map[schema.GroupVersionKind][]client.Selector
```

Reply:
Be that as it may, but the UX of "head over to the client package to get a selector to use in the cache package" is plain bad. So is the UX of "and then hopefully your IDE is smart in suggesting the right things, otherwise tough luck".

Cache selectors belong in the cache package, not in the client package. The fact that there is some overlap doesn't change anything about that.

Reply:
fair enough, let me put in place something with similar functionality but withing "pkg/cache".

Reply:
Added new commit to move the selector and friends from "pkg/client/options.go" to "pkg/cache/selector.go" https://github.com/kubernetes-sigs/controller-runtime/pull/1435/commits/c9435b0a8ab0a68912916336f50f724a227c0bd3

Now users only need to reference "cache" package.